Convert Flask templates to React components

omarkmu opened the pull request:
Converted HTML pages to React components. Also, added proper display and validation for input elements on each page. This PR does not change the `App` component other than an updated import and the global CSS changes. That component will need to be updated with the new components in a different PR.

### Changes
- Converted Login, Registration, and Profile pages to React components
- Extracted reusable HTML to components
- Added validation and display of saved data for all input elements
- Improved backend validation
- Modified React file structure; user-visible pages are in `src/pages`, components are in `src/components`, and everything else remains in the `src` folder
- Renamed `index` route to `app`
- Updated app styling (slightly)
- Various refactoring changes and bugfixes

### Components
Since this PR adds a lot of components and significantly modifies the structure of the project, I thought it'd be best to add a quick summary of the new components. I can add more detail via doc comments in the code if requested.

- **AnchorButton**: Renders a button which acts as link (on click, it opens the specified path). Accepts `href` and `text`.
- **Container**: Renders a div with the "container" class. Can have children.
- **Flashes**: Renders flashes from flask in a `Row`. Accepts `flashes` (array of strings).
- **Form**: Renders a form. Accepts `method` (`'POST'` by default), `action`, and `token` (the CSRF token provided by Flask; see `Main.js`).
- **Input**: Renders an input element with validation handling. Accepts:
    - `id`: a string which is assigned to both the `id` and `name` attributes.
    - `value`: the default value of the input field.
    - `form`: The form data from flask (see `Main.js`).
    - any attributes which are valid on `input`s.
- **Row**: Renders a div with the "container-row" class. Can have children. Accepts `marginFactor` (multiplicative factor to apply to `margin-bottom` in the CSS).
- **Select**: Renders a select element with options. Accepts:
    - `id`: a string which is assigned to both the `id` and `name` attributes.
    - `selected`: a string which specifies the value of the option which should be selected by default.
    - `values`: an array of string values which will be rendered as options.
    - `labels`: an array of string values which will be used as option text (if omitted, the values are used as text).
    - any attributes which are valid on `select`s.
- **Submit**: Renders a button which acts as a form submission button. **This is not a button of type "submit".** The component manually handles the click event so that form validation does not break on attempted submission.

omarkmu committed:
chore: Converted login to a React component
refactor: Added render and redirect helper functions to app.py
fix: Fixed BAD_REQUEST bug with search endpoint

omarkmu committed:
chore: Converted registration to a React component
refactor: Removed unused methods from routes
refactor: Added default values and form errors in template
fix: Fixed incorrect required value in input fields
fix: Removed invalid font attribute from CSS

omarkmu committed:
chore: Converted profile to a React component

feat: Added display for saved weight and birthdate on profile page

refactor: Removed unused arguments in forms.py
refactor: Added early return for GET requests where appropriate

fix: Fixed missing height inputs on profile page
fix: Removed format specifier from birthdate field in Profile form

omarkmu committed:
chore: Extracted reusable HTML to components
chore: Added validation for birth date
fix: Fixed missing profile values after submission
fix: Fixed false positive height_inches validation error

Also added an exclusion for a linter rule, to allow the ...rest syntax.

omarkmu committed:
feat: Added validation logic to input elements
chore: Improved backend form validation
refactor: Switched to early returns for route functions
refactor: Removed specific Input type components
refactor: Merged InputComponents into HTMLComponents
refactor: Renamed index page to app

omarkmu committed:
chore: Added Row component for structure
chore: Improved AnchorButton
refactor: Renamed LinkButton to AnchorButton
refactor: Moved HTMLComponents to src directory
fix: Removed register text from profile page

omarkmu committed:
refactor: Improved React file structure
refactor: Moved AnchorButton to HTMLComponents

omarkmu committed:
chore: Added form data persistence
style: Wrapped input labels in label elements
style: Improved CSS
fix: Fixed a bug in Profile.json property

omarkmu committed:
refactor: Renamed AppRouter to Main
refactor: Change Option to non-exported component
chore: Fixed potential issue in Input component
chore: Changed Row to div rather than span

omarkmu committed:
fix: Fixed bug with saving decimal weights
fix: Fixed null coalesce checks
refactor: Moved default form logic to template

BrandonDwyer commented on app.py:
```diff
@@ -3,38 +3,49 @@
-    mock_data = {"your": "data here"}
-    data = json.dumps(mock_data)
-    return flask.render_template("index.html", data=data)
+    # NOTE: left this here because we may need it for favorites
+    data = {"your": "data here"}
+    return render(data=data)
```
Good thinking leaving this here if we are able to complete this stretch feature.

BrandonDwyer reviewed (commented):
Overall these changes look pretty good to me and offer a more cohesive product going forward with being fully reactJS instead of a combo of HTML and reactJS. I would wait to merge these changes until another team member reviews them just to make sure I did not miss anything due to how many files have been changed.

dhartman5 commented on app.py:
```diff
@@ -3,38 +3,49 @@
 
 
-@app.route("/index", methods=["GET", "POST"])
+@app.route("/app")
```
Will these not have methods?

omarkmu replied:
The GET method is the default if the method is not specified, and we aren't using the POST method on that endpoint (the frontend uses `fetch` with the `/api/search` endpoint).

dhartman5 replied:
👍

dhartman5 commented on app.py:
```diff
@@ -43,69 +54,65 @@ def profile():
+
+    # delete existing profile (loop just in case)
+    for existing_profile in Profile.query.filter_by(user_id=current_user.id):
+        db.session.delete(existing_profile)
```
What're we deleting the profile information for? 

Is this when the user wants to enter in new info?

omarkmu replied:
A user should have a single associated `Profile`, so this is ensuring that the profile is unique before updating it. This could probably be done in a cleaner way using whatever SQLAlchemy's form of `UPDATE` is.

Honestly, considering the fact that it's a one-to-one relationship, it would probably be simpler to move everything from the `Profile` table to `BTUser`. That being said, this approach works too. "If it ain't broke, don't fix it", I suppose.

dhartman5 replied:
Oh ok gotcha. That's true, I think it's fine having them separated due to password safety since its stored in BTUser and not Profile.

dhartman5 commented on app.py:
```diff
@@ -43,69 +54,65 @@ def profile():
+    height_feet = form.height_feet.data or 0
+    height_inches = form.height_inches.data or 0
+
+    updated_profile = Profile(
```
Thank you for adding this!

dhartman5 commented on app.py:
```diff
@@ -43,69 +54,65 @@ def profile():
-    elif len(form.password.data) < 4:
-        flash("Please enter a password greater than 4 characters")
-    elif len(form.password.data) > 20:
-        flash("Please enter a password shorter than 20 characters")
```
Are we adding new checks? We don't want the user to be able to input infinite length usernames.

omarkmu replied:
I moved the logic for all of the validation to `forms.py`. This way, rather than seeing the message as a flash below the input, error messages can be displayed natively on the individual input field with `setCustomValidity`.

dhartman5 replied:
Oh sweet! That's perfect!

omarkmu committed:
fix: Fixed required field validation logic
fix: Fixed bug that prevented display of inches when inches is 0
chore: Added options for default values in Select
chore: Added validation for profile selects
chore: Removed exception for date type in Input
